### PR TITLE
feat(rubicon): Add Rubicon.finance and Rubicon Bath Tokens

### PR DESCRIPTION
**Pull request by bghughes**
## Description

This is an initial PR to initialize [Rubicon](https://www.rubicon.finance/) as a Zapper project. We are excited to integrate with Zapper and this PR represents a basic implementation that tracks [Bath Tokens](https://docs.rubicon.finance/rubicon-docs/contracts/rubicon-pools/bath-token) and their underlying value; the same way we do in our [app](https://app.rubicon.finance/).

## Checklist

- [X] I have followed the [Contributing Guidelines](../CONTRIBUTING.md)
- [X] (optional) As a contributor, my Ethereum address/ENS is: `bghughes.eth`
- [X] (optional) As a contributor, my Twitter handle is: `@_benjaminhughes`

## How to test?

A sanity check on the ERC-20 Bath Token balances can be performed via Etherscan as all contracts are verified on-chain and audited. Please see a list of all the mainnet Bath Tokens [here](https://docs.rubicon.finance/rubicon-docs/contracts/deployments#optimism-mainnet-rubicon-v1).

## Future Works
The storing of Bath Token addresses will be overhauled to be dynamic at the protocol level (and here when fetching as presently a hardcoded string array is used) in future updates. Moreover, in the next PR, we will incorporate custom logos and more information to improve the UX.

Please let me know if I can help the process or provide any additional information - we are pumped to be on Zapper!

### CODE REVIEW

File: src/apps/rubicon/optimism/rubicon.bathToken.token-fetcher.ts [hunk 1]
```diff
@@ -0,0 +1,124 @@
...
+          address: btAddress,
+          network,
+          symbol,
+          decimals,
+          supply,
+          tokens,
+          price,
+          pricePerShare,
+          // TODO: add any data props here
+          dataProps: {},
```
Comment (immasandwich) [inline src/apps/rubicon/optimism/rubicon.bathToken.token-fetcher.ts:111]: No data props required, so you can simply remove the comment.

File: src/apps/rubicon/optimism/rubicon.bathToken.token-fetcher.ts [hunk 2]
```diff
@@ -0,0 +1,124 @@
...
+          displayProps: {
+            label: '',
+            images: [],
```
Comment (immasandwich) [inline src/apps/rubicon/optimism/rubicon.bathToken.token-fetcher.ts:114]: For the label, use `getLabelFromToken(underlyingToken)`, and for images use `getImagesFromToken(underlyingToken)`. It'd be a good idea to use `buildDollarDisplayItem(underlyingToken.price)` for the `secondaryLabel` as well.

For more info, see [here](https://studio.zapper.fi/docs/concepts/app-tokens)

File: src/apps/rubicon/optimism/rubicon.bathToken.token-fetcher.ts [hunk 3]
```diff
@@ -0,0 +1,124 @@
...
+          },
+        };
+
+        return token;
+      }),
+    );
+
+    return tokens;
```
Comment (immasandwich) [inline src/apps/rubicon/optimism/rubicon.bathToken.token-fetcher.ts:122]: Remember to filter out `null` values from the above `Promise.all` with `_.compact(tokens)`

**Review by immasandwich — COMMENTED** [review]
Looks quite good sers! Great first contribution. A few comments to fix, then we can ship it. Please also add a logo to `src/apps/rubicon/assets/logo.png`

File: src/apps/rubicon/optimism/rubicon.bathToken.token-fetcher.ts [hunk 4]
```diff
@@ -0,0 +1,124 @@
...
+  async getPositions() {
+    // For now, hardcoded in docs
+    const bathTokens = [
+      '0xB0bE5d911E3BD4Ee2A8706cF1fAc8d767A550497',
+      '0x7571CC9895D8E997853B1e0A1521eBd8481aa186',
+      '0xe0e112e8f33d3f437D1F895cbb1A456836125952',
+      '0x60daEC2Fc9d2e0de0577A5C708BcaDBA1458A833',
+      '0xfFBD695bf246c514110f5DAe3Fa88B8c2f42c411',
+      '0xeb5F29AfaaA3f44eca8559c3e8173003060e919f',
+      '0x574a21fE5ea9666DbCA804C9d69d8Caf21d5322b',
```
Comment (immasandwich) [inline src/apps/rubicon/optimism/rubicon.bathToken.token-fetcher.ts:37]: Lowercase the addresses here, and remove any comments you feel are unnecessary below.

File: src/apps/rubicon/optimism/rubicon.bathToken.token-fetcher.ts (hunk 2 above)
Comment (bghughes) [reply src/apps/rubicon/optimism/rubicon.bathToken.token-fetcher.ts:114]: Great, thank you, resolved in my newest commit. Moreover we have custom BathToken logos I would like to include in a future update - for now we are working on setting up a CDN to serve this.

File: src/apps/rubicon/optimism/rubicon.bathToken.token-fetcher.ts (hunk 4 above)
Comment (bghughes) [reply src/apps/rubicon/optimism/rubicon.bathToken.token-fetcher.ts]: `    const bathTokens = [
      '0xB0bE5d911E3BD4Ee2A8706cF1fAc8d767A550497'.toLowerCase(),
      '0x7571CC9895D8E997853B1e0A1521eBd8481aa186'.toLowerCase(),
      '0xe0e112e8f33d3f437D1F895cbb1A456836125952'.toLowerCase(),
      '0x60daEC2Fc9d2e0de0577A5C708BcaDBA1458A833'.toLowerCase(),
      '0xfFBD695bf246c514110f5DAe3Fa88B8c2f42c411'.toLowerCase(),
      '0xeb5F29AfaaA3f44eca8559c3e8173003060e919f'.toLowerCase(),
      '0x574a21fE5ea9666DbCA804C9d69d8Caf21d5322b'.toLowerCase(),
    ];
 `
 
 Resolved with this pattern.

**Comment by bghughes** [comment]
> Looks quite good sers! Great first contribution. A few comments to fix, then we can ship it. Please also add a logo to `src/apps/rubicon/assets/logo.png`

Thank you so much for the swift review @immasandwich. I have resolved all of your comments and added our logo in [d56197a](https://github.com/Zapper-fi/studio/pull/922/commits/d56197a6d67884a667394070cf3d4e8519cff616).

Please let me know what else I can do to be helpful or what we should do to accelerate the process. Thanks again 🚀

**Review by immasandwich — COMMENTED** [review]
One more thing, can you rename `rubicon.bathToken.token-fetcher.ts` to `rubicon.bath-token.token-fetcher.ts`? We specifically kebab-case the group in file names.

File: src/apps/rubicon/optimism/rubicon.bathToken.token-fetcher.ts [hunk 5]
```diff
@@ -0,0 +1,132 @@
...
+  async getPositions() {
+    // For now, hardcoded in docs
+    const bathTokens = [
+      '0xB0bE5d911E3BD4Ee2A8706cF1fAc8d767A550497'.toLowerCase(),
+      '0x7571CC9895D8E997853B1e0A1521eBd8481aa186'.toLowerCase(),
+      '0xe0e112e8f33d3f437D1F895cbb1A456836125952'.toLowerCase(),
+      '0x60daEC2Fc9d2e0de0577A5C708BcaDBA1458A833'.toLowerCase(),
+      '0xfFBD695bf246c514110f5DAe3Fa88B8c2f42c411'.toLowerCase(),
+      '0xeb5F29AfaaA3f44eca8559c3e8173003060e919f'.toLowerCase(),
+      '0x574a21fE5ea9666DbCA804C9d69d8Caf21d5322b'.toLowerCase(),
```
Comment (immasandwich) [inline src/apps/rubicon/optimism/rubicon.bathToken.token-fetcher.ts:40]: Don't call `toLowerCase()`, actually lowercase these addresses (VSCode has as plugin lol, I use it frequently). We have also recently introduced an ESLint rule that automatically should lowercase addresses in the codebase (integrated today, so if you merge in the latest from master, you'll be able to run this rule to auto-fix lowercasing).

**Comment by bghughes** [comment]
> One more thing, can you rename `rubicon.bathToken.token-fetcher.ts` to `rubicon.bath-token.token-fetcher.ts`? We specifically kebab-case the group in file names.

No problem! Resolved in [ae10227](https://github.com/Zapper-fi/studio/pull/922/commits/ae10227aed12d984d28fea9b2838ed1de5831e1b) - please lmk if there's anything else I can do to be helpful 😄

**Review by immasandwich — APPROVED** [review]

**Comment by bghughes** [comment]
Thank you @immasandwich, quick question: can we use custom logos for Bath Tokens on the platform? For example, if we provided custom logos like those found here, could they populate in Zapper? It seems most projects do not do this so just wanted to ask.

Thank you again and we will have a follow-up PR for you that improves our implementation to a more scalable state, with a capitalized "r" in Rubicon.

**Pull request merged**